### ISSUE TITLE
-Werror=restrict with gcc-12 -std=c++2{0,3}

### ISSUE THREAD

**Issue by hainest** [dyninst/dyninst#1422]
From https://github.com/dyninst/dyninst/actions/runs/4836794898:

```
error: 'void* __builtin_memcpy(void*, const void*, long unsigned int)' accessing 9223372036854775810
or more bytes at offsets [2, 9223372036854775807] and 1 may overlap up to 9223372036854775813 
bytes at offset -3 [-Werror=restrict])
```

Code:

```
virtual std::string format(formatStyle) const {
  std::string retVal;
#if defined(DEBUG_MEMORY_ACCESS_WIDTH)
...
#endif
  retVal += "[" + addressToDereference->format() + "]";     <---------
  return retVal;
}
```

@kupsch This feels like a compiler bug. Thoughts?

Full error message:

```
[  9%] Building CXX object instructionAPI/CMakeFiles/instructionAPI.dir/src/Instruction.C.o
In file included from /usr/include/c++/12/string:40,
                 from /dyninst/src/common/h/util.h:204,
                 from /dyninst/src/common/src/Types.h:94,
                 from /dyninst/src/instructionAPI/src/Instruction.C:33:
In static member function 'static constexpr std::char_traits<char>::char_type* std::char_traits<char>::copy(char_type*, const char_type*, std::size_t)',
    inlined from 'static constexpr void std::__cxx11::basic_string<_CharT, _Traits, _Alloc>::_S_copy(_CharT*, const _CharT*, size_type) [with _CharT = char; _Traits = std::char_traits<char>; _Alloc = std::allocator<char>]' at /usr/include/c++/12/bits/basic_string.h:423:21,
    inlined from 'constexpr std::__cxx11::basic_string<_CharT, _Traits, _Allocator>& std::__cxx11::basic_string<_CharT, _Traits, _Alloc>::_M_replace(size_type, size_type, const _CharT*, size_type) [with _CharT = char; _Traits = std::char_traits<char>; _Alloc = std::allocator<char>]' at /usr/include/c++/12/bits/basic_string.tcc:532:22,
    inlined from 'constexpr std::__cxx11::basic_string<_CharT, _Traits, _Alloc>& std::__cxx11::basic_string<_CharT, _Traits, _Alloc>::replace(size_type, size_type, const _CharT*, size_type) [with _CharT = char; _Traits = std::char_traits<char>; _Alloc = std::allocator<char>]' at /usr/include/c++/12/bits/basic_string.h:2171:19,
    inlined from 'constexpr std::__cxx11::basic_string<_CharT, _Traits, _Alloc>& std::__cxx11::basic_string<_CharT, _Traits, _Alloc>::insert(size_type, const _CharT*) [with _CharT = char; _Traits = std::char_traits<char>; _Alloc = std::allocator<char>]' at /usr/include/c++/12/bits/basic_string.h:1928:22,
    inlined from 'constexpr std::__cxx11::basic_string<_CharT, _Traits, _Allocator> std::operator+(const _CharT*, __cxx11::basic_string<_CharT, _Traits, _Allocator>&&) [with _CharT = char; _Traits = char_traits<char>; _Alloc = allocator<char>]' at /usr/include/c++/12/bits/basic_string.h:3541:36,
    inlined from 'virtual std::string Dyninst::InstructionAPI::Dereference::format(Dyninst::InstructionAPI::formatStyle) const' at /dyninst/src/instructionAPI/h/Dereference.h:153:19:
/usr/include/c++/12/bits/char_traits.h:431:56: error: 'void* __builtin_memcpy(void*, const void*, long unsigned int)' accessing 9223372036854775810 or more bytes at offsets [2, 9223372036854775807] and 1 may overlap up to 9223372036854775813 bytes at offset -3 [-Werror=restrict]
  431 |         return static_cast<char_type*>(__builtin_memcpy(__s1, __s2, __n));
      |                                        ~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~
cc1plus: all warnings being treated as errors
gmake[2]: *** [instructionAPI/CMakeFiles/instructionAPI.dir/build.make:76: instructionAPI/CMakeFiles/instructionAPI.dir/src/Instruction.C.o] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:413: instructionAPI/CMakeFiles/instructionAPI.dir/all] Error 2
gmake[1]: *** Waiting for unfinished jobs....
```

**Comment by kupsch** [comment]
Yes, I'd say it's a compiler bug.  Without `-Werror` there are 14 of these. They all involve a `std::string` operation that has a `char *` literal argument.  There are 3 at lines 154, 161, and 162 of the files *parseAPI/src/ProbabilisticParser.C* below. Not clear why 156 didn't trigger the problem, but most the c-string literals that trigger the warning are small (1 maybe 2 bytes plus the NULL). Most of the warnings are of the form of lines 161 and 162 where the string literal is passed as the lhs and a std::string as the rhs to `std::string::operator+`.

```
151 string IdiomTerm::human_format() const {
152     string entryname;
153     if(*this == WILDCARD_TERM)
154         entryname = "*";
155     else if(entryNames_IAPI.find((entryID)(entry_id)) == entryNames_IAPI.end()    ) { 
156         entryname = "[UNMAPED]";
157         fprintf(stderr, "Found entryID not mapped in entryNames_IAPI %d\n", en    try_id);
158     }
159     else {
160         entryname = entryNames_IAPI[(entryID)(entry_id)];
161         if (arg1 != NOARG) entryname += " " + HandleAnOperand(arg1, 1);
162         if (arg2 != NOARG) entryname += "," + HandleAnOperand(arg2, 1);
163     }
164 
165     return entryname;
166 
167 }
```

**Comment by kupsch** [comment]
RELEASE build on fedora 38 using g++ version 13 and C++20 builds without warning.

**Comment by hainest** [comment]
Fixed by #1664
